src/inc/til/rle.h

```diff
@@ -0,0 +1,1064 @@
+            _total_length(length)
+        {
+            if (length)
+            {
+                _runs.emplace_back(value, length);
+            }
+        }
+
+        void swap(basic_rle& other) noexcept
+        {
+            _runs.swap(other._runs);
+            std::swap(_total_length, other._total_length);
+        }
+
+        bool empty() const noexcept
+        {
+            return _total_length == 0;
+        }
+
+        // Returns the total length of all runs as encoded.
+        size_type size() const noexcept
+        {
+            return _total_length;
+        }
+
+        // This method gives access to the raw run length encoded array
+        // and allows users of this class to iterate over those.
+        const container& runs() const noexcept
+        {
+            return _runs;
+        }
+
+        // Get the value at the position
+        const_reference at(size_type position) const
+        {
+            const auto begin = _runs.begin();
+            const auto end = _runs.end();
+
+            rle_scanner scanner(begin, end);
+            auto it = scanner.scan(position).first;
+
+            if (it == end)
+            {
+                throw std::out_of_range("position out of range");
+            }
+
+            return it->value;
+        }
+
+        [[nodiscard]] basic_rle slice(size_type start_index, size_type end_index) const noexcept
+        {
+            if (end_index > _total_length)
+            {
+                end_index = _total_length;
+            }
+
+            if (start_index >= end_index)
+            {
+                return {};
+            }
+
+            // Thanks to the prior conditions we can safely assume that:
+            // * 0 <= start_index < _total_length
+            // * 0 < end_index <= _total_length
+            // * start_index < end_index
+            //
+            // --> It's safe to subtract 1 from end_index
+
+            rle_scanner scanner(_runs.begin(), _runs.end());
+            auto [begin_run, start_run_pos] = scanner.scan(start_index);
+            auto [end_run, end_run_pos] = scanner.scan(end_index - 1);
+
+            container slice{ begin_run, end_run + 1 };
+            slice.back().length = end_run_pos + 1;
+            slice.front().length -= start_run_pos;
+
+            return { std::move(slice), static_cast<size_type>(end_index - start_index) };
+        }
+
+        // Set the range [start_index, end_index) to the given value.
+        void replace(size_type start_index, size_type end_index, const value_type& value)
+        {
+            _check_indices(start_index, end_index);
+
+            const rle_type replacement{ value, static_cast<size_type>(end_index - start_index) };
+            _replace_unchecked(start_index, end_index, { &replacement, 1 });
+        }
+
+        // Replace the range [start_index, end_index) with the given run.
+        // NOTE: This can change the size/length of the vector.

```

Comment:
```suggestion
        // NOTE: This can change the size/length of the container.

```

You made the container potentially generic, just defaulting as vector.